File: typescript/review/permutations.ts
function permutations<T>(elements: T[]): T[][] {
  if (elements.length === 0) return [[]];

  const firstEl = elements[0];
  const withoutFirst = elements.slice(1);
  const permsWithoutFirst = permutations(withoutFirst);
  const allPermutations: T[][] = [];

  for (let perm of permsWithoutFirst) {
    for (let i = 0; i <= perm.length; i++) {
      const permWithFirst = [...perm.slice(0, i), firstEl, ...perm.slice(i)];
      allPermutations.push(permWithFirst);
    }
  }

  return allPermutations;
}

console.log(permutations([1, 2, 'a']));

File: python/today2/permutations.py
# https://leetcode.com/problems/permutations/
from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return [[]]

        firstEl = nums[0]
        withoutFirst = nums[1:]
        permsWithoutFirst = self.permute(withoutFirst)
        allPermutations = []

        for perm in permsWithoutFirst:
            i = 0
            while i <= len(perm):
                permWithFirst = perm[0:i] + [firstEl] + perm[i:]
                allPermutations.append(permWithFirst)
                i += 1
        
        return allPermutations
